File: testResources/integration_test_project/frontend/javascript/tests/util/type_check.js



export function isRawGtv(value) {
    return (
        value === null ||
        typeof value === 'boolean' ||
        value instanceof Buffer ||
        typeof value === 'string' ||
        typeof value === 'number' ||
        typeof value === 'bigint' ||
        (Array.isArray(value) && value.every(isRawGtv)) ||
        (value !== null && typeof value === 'object' && 'key' in value && 'value' in value)
    );
}

export function isMapWithTestStructAsKey(response) {
    return (
        Array.isArray(response) && 
        response.every(
            item => 
                Array.isArray(item) && 
                item.length === 2 && 
                typeof item[0] === 'object' && 
                'a' in item[0] && // assuming TestStruct has an 'a' property
                typeof item[1] === 'string'
        )
    );
}

File: testResources/integration_test_project/frontend/python/tests/util/type_check.py
from typing import Any


def is_raw_gtv(value: Any) -> bool:
    if value is None:
        return True

    if isinstance(value, (int, str, bytes, bool)):
        return True

    if isinstance(value, list):
        return all(is_raw_gtv(item) for item in value)

    if isinstance(value, dict):
        return all(
            isinstance(key, str) and is_raw_gtv(val) for key, val in value.items()
        )

    return False


def is_test_enum(value: Any) -> bool:
    return isinstance(value, str) and value in ["a"]


def is_map_with_test_struct_as_key(value: Any) -> bool:
    if not isinstance(value, list):
        return False

    for item in value:
        if not isinstance(item, list) or len(item) != 2:
            return False

        key, _ = item
        if not isinstance(key, list) or len(key) != 1:
            return False

        if not isinstance(key[0], bool):
            return False

    return True

